hubspot-style/src/main/java/com/hubspot/immutables/utils/WireSafeEnum.java

```diff
@@ -0,0 +1,110 @@
+package com.hubspot.immutables.utils;
+
+import java.io.IOException;
+import java.util.Optional;
+
+import com.fasterxml.jackson.annotation.JsonValue;
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.core.JsonToken;
+import com.fasterxml.jackson.databind.BeanProperty;
+import com.fasterxml.jackson.databind.DeserializationContext;
+import com.fasterxml.jackson.databind.JavaType;
+import com.fasterxml.jackson.databind.JsonDeserializer;
+import com.fasterxml.jackson.databind.JsonMappingException;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.fasterxml.jackson.databind.deser.ContextualDeserializer;
+import com.hubspot.immutables.utils.WireSafeEnum.Deserializer;
+
+@JsonDeserialize(using = Deserializer.class)
+public final class WireSafeEnum<T extends Enum<T>> {
+  private final Class<T> enumType;
+  private final String stringValue;
+  private final Optional<T> enumValue;
+
+  private WireSafeEnum(Class<T> enumType, String stringValue) {
+    this.enumType = enumType;
+    this.stringValue = stringValue;
+    this.enumValue = tryParse(enumType, stringValue);
+  }
+
+  @SuppressWarnings("unchecked")
+  private WireSafeEnum(T value) {
+    this.enumType = (Class<T>) value.getClass();
+    this.stringValue = value.name();
+    this.enumValue = Optional.of(value);
+  }
+
+  public static <T extends Enum<T>> WireSafeEnum<T> of(T value) {
+    return new WireSafeEnum<>(value);
```

Comment:
I realize this is a POC, but if we actually roll this out, I think instances created from this method should be reused.

Reply:
updated